src/device-state/target-state.ts

```diff
@@ -0,0 +1,157 @@
+import type { TargetState } from '../types/state';
+import { EventEmitter } from 'events';
+import { InternalInconsistencyError } from '../lib/errors';
+import { getRequestInstance } from '../lib/request';
+import * as url from 'url';
+import * as config from '../config';
+import type { Headers } from 'request';
+import { delay } from 'bluebird';
+import constants = require('../lib/constants');
+import * as _ from 'lodash';
+import { writeLock } from '../lib/update-lock';
+import Bluebird = require('bluebird');
+import type StrictEventEmitter from 'strict-event-emitter-types';
+
+interface TargetStateEvents {
+	'target-state-update': (
+		targetState: TargetState,
+		force: boolean,
+		isFromApi: boolean,
+	) => void;
+}
+export const emitter: StrictEventEmitter<
+	EventEmitter,
+	TargetStateEvents
+> = new EventEmitter();
+
+const lockGetTarget = () =>
+	writeLock('getTarget').disposer((release) => release());
+
+let cache: {
+	etag?: string | string[];
+	body: TargetState;
+};
+
+/**
+ * The last fetch attempt, success or fail
+ */
+export let lastFetch: ReturnType<typeof process.hrtime>;
+/**
+ * Attempts to update the target state
+ * @param force Emitted with the 'target-state-update' event update as necessary
+ * @param isFromApi Emitted with the 'target-state-update' event update as necessary
+ */
+export const update = async (
+	// TODO: Is there a better way than passing these params here just to emit them if there is an update?
+	force = false,
+	isFromApi = false,
+): Promise<void> => {
+	return Bluebird.using(lockGetTarget(), async () => {
+		const { uuid, apiEndpoint, apiTimeout } = await config.getMany([
+			'uuid',
+			'apiEndpoint',
+			'apiTimeout',
+		]);
+
+		if (typeof apiEndpoint !== 'string') {
+			throw new InternalInconsistencyError(
+				'Non-string apiEndpoint passed to ApiBinder.getTargetState',
+			);
+		}
+
+		const endpoint = url.resolve(apiEndpoint, `/device/v2/${uuid}/state`);
+		const request = await getRequestInstance();
+
+		const params: Headers = {
+			json: true,
+		};
+
+		if (typeof cache?.etag === 'string') {
+			params.headers = {
+				'If-None-Match': cache.etag,
+			};
+		}
+
+		const [{ statusCode, headers }, body] = await request
+			.getAsync(endpoint, {
+				json: true,
+			})
+			.timeout(apiTimeout);
+
+		if (statusCode === 304) {
+			// There's no change so no need to update the cache or emit a change event
+			return;
+		}
+
+		cache = {
+			etag: headers.etag,
+			body,
+		};
+
+		emitter.emit('target-state-update', _.cloneDeep(body), force, isFromApi);
+	}).finally(() => {
+		lastFetch = process.hrtime();
+	});
+};
+
+let targetStateFetchErrors = 0;
+const poll = async (skipFirstGet: boolean = false): Promise<void> => {
+	try {
+		// TODO: I'd prefer to get this only once and then add an event listener for updates
+		const appUpdatePollInterval = await config.get('appUpdatePollInterval');
```

Comment:
This should definitely be handled before merge.

Reply:
Currently, if poll cannot get `appUpdatePollInterval` it will wait 10 seconds and then recursively call itself. Getting this config value is the first thing the function does so effectively poll will keep trying to read the config every 10 seconds until successful then continue.

What I have implemented maintains this exact behavior but provides the benefit of best case only querying the config once. I did this by moving all (1) config queries to startPoll. This function already had a try catch for getting `instantUpdates` which would default to a value if there was an error. The default value is basically useless though since the poll function will be stuck in a loop trying to read `appUpdatePollInterval` from config.